File: bindings/jvm/src/test/java/com/goplasmatic/datalogic/EngineTest.java
/* SPDX-License-Identifier: Apache-2.0 */
package com.goplasmatic.datalogic;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class EngineTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    @Test
    void version_is_non_empty() {
        assertFalse(Engine.version().isEmpty());
    }

    @Test
    void apply_one_shot_returns_json_result() {
        try (Engine engine = new Engine()) {
            assertEquals("3", engine.apply("{\"+\":[1,2]}", "{}"));
        }
    }

    @Test
    void compile_then_evaluate_reuses_rule() {
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"var\":\"x\"}")) {
            for (int x : new int[]{1, 7, 42}) {
                assertEquals(String.valueOf(x), rule.evaluate("{\"x\":" + x + "}"));
            }
        }
    }

    @Test
    void session_reuses_arena_across_calls() {
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"*\":[{\"var\":\"x\"},2]}");
             Session session = engine.openSession()) {
            for (int x : new int[]{3, 5, 8}) {
                assertEquals(String.valueOf(x * 2), session.evaluate(rule, "{\"x\":" + x + "}"));
            }
            assertTrue(session.allocatedBytes() > 0);
        }
    }

    @Test
    void parse_error_throws_ParseException() {
        try (Engine engine = new Engine()) {
            ParseException ex = assertThrows(ParseException.class, () -> engine.compile("not-json{{"));
            assertEquals("ParseError", ex.errorType());
            assertNotNull(ex.getMessage());
        }
    }

    @Test
    void evaluate_error_throws_with_operator_and_path() {
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"throw\":\"boom\"}")) {
            EvaluateException ex = assertThrows(EvaluateException.class, () -> rule.evaluate("{}"));
            assertEquals("Thrown", ex.errorType());
            assertNotNull(ex.pathJson());
            assertTrue(ex.pathJson().startsWith("["));
        }
    }

    @Test
    void templating_engine_constructs() {
        try (Engine engine = new Engine(true)) {
            assertNotNull(engine.apply("{\"+\":[1,1]}", "{}"));
        }
    }

    @Test
    void flagd_sem_ver_operator_is_available() {
        try (Engine engine = new Engine()) {
            assertEquals("true", engine.apply("{\"sem_ver\":[\"1.2.3\",\"<\",\"2.0.0\"]}", "{}"));
        }
    }

    @Test
    void traced_session_returns_result_and_steps() throws Exception {
        try (Engine engine = new Engine();
             TracedSession session = engine.openTracedSession()) {
            TracedRun run = session.evaluate("{\"+\":[{\"var\":\"x\"},1]}", "{\"x\":41}");
            assertTrue(run.isSuccess());
            assertEquals(42, run.result().asInt());
            assertTrue(run.steps().isArray() && run.steps().size() > 0);
            assertTrue(run.expressionTree().isObject());
            assertNull(run.error());
        }
    }

    @Test
    void traced_session_surfaces_runtime_error_in_payload() {
        try (Engine engine = new Engine();
             TracedSession session = engine.openTracedSession()) {
            TracedRun run = session.evaluate("{\"throw\":\"boom\"}", "{}");
            assertFalse(run.isSuccess());
            assertNotNull(run.error());
            assertTrue(run.structuredError().isObject());
        }
    }

    @Test
    void builder_registers_custom_operator() {
        try (Engine engine = Engine.builder()
                .addOperator("double", argsJson -> {
                    JsonNode arr = MAPPER.readTree(argsJson);
                    return String.valueOf(arr.get(0).asInt() * 2);
                })
                .build()) {
            assertEquals("42", engine.apply("{\"double\":[21]}", "{}"));
        }
    }

    @Test
    void builder_custom_operator_error_propagates() {
        try (Engine engine = Engine.builder()
                .addOperator("boom", argsJson -> { throw new RuntimeException("custom-failure"); })
                .build()) {
            EvaluateException ex = assertThrows(EvaluateException.class,
                    () -> engine.apply("{\"boom\":[]}", "{}"));
            assertTrue(ex.getMessage().contains("custom-failure"), "got: " + ex.getMessage());
        }
    }

    @Test
    void builder_set_config_json_strict_preset_takes_effect() {
        // Default config: null coerces to 0 and the sum evaluates.
        try (Engine engine = new Engine()) {
            assertEquals("1", engine.apply("{\"+\":[null,1]}", "{}"));
        }
        // Strict preset: the same rule rejects the non-numeric null.
        try (Engine engine = Engine.builder()
                .setConfigJson("{\"preset\":\"strict\"}")
                .build()) {
            assertThrows(EvaluateException.class, () -> engine.apply("{\"+\":[null,1]}", "{}"));
        }
    }

    @Test
    void builder_set_config_json_rejects_bad_input() {
        // Malformed JSON surfaces the parser's message.
        EvaluateException malformed = assertThrows(EvaluateException.class,
                () -> Engine.builder().setConfigJson("not-json{{"));
        assertEquals("ConfigurationError", malformed.errorType());
        assertNotNull(malformed.getMessage());
        assertFalse(malformed.getMessage().isEmpty());

        // Unknown enum values fail loudly instead of being ignored.
        EvaluateException bogus = assertThrows(EvaluateException.class,
                () -> Engine.builder().setConfigJson("{\"preset\":\"bogus\"}"));
        assertTrue(bogus.getMessage().contains("bogus"), "got: " + bogus.getMessage());
    }

    @Test
    void builder_set_config_json_chains_with_templating() {
        try (Engine engine = Engine.builder()
                .withTemplating(true)
                .setConfigJson("{\"preset\":\"strict\"}")
                .build()) {
            assertEquals("3", engine.apply("{\"+\":[1,2]}", "{}"));
            assertThrows(EvaluateException.class, () -> engine.apply("{\"+\":[null,1]}", "{}"));
        }
    }
}

File: bindings/php/tests/EngineTest.php
<?php

declare(strict_types=1);

/* SPDX-License-Identifier: Apache-2.0 */

namespace Goplasmatic\Datalogic\Tests;

use Goplasmatic\Datalogic\Engine;
use Goplasmatic\Datalogic\Exception\DatalogicException;
use Goplasmatic\Datalogic\Exception\EvaluateException;
use Goplasmatic\Datalogic\Exception\ParseException;
use PHPUnit\Framework\TestCase;

final class EngineTest extends TestCase
{
    public function test_version_is_non_empty(): void
    {
        self::assertNotSame('', Engine::version());
    }

    public function test_apply_one_shot_returns_json_result(): void
    {
        $engine = new Engine();
        self::assertSame('3', $engine->apply('{"+":[1,2]}', '{}'));
    }

    public function test_compile_then_evaluate_reuses_rule(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"var":"x"}');
        foreach ([1, 7, 42] as $x) {
            self::assertSame((string) $x, $rule->evaluate('{"x":' . $x . '}'));
        }
    }

    public function test_session_reuses_arena_across_calls(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"*":[{"var":"x"},2]}');
        $session = $engine->openSession();
        foreach ([3, 5, 8] as $x) {
            self::assertSame((string) ($x * 2), $session->evaluate($rule, '{"x":' . $x . '}'));
        }
        self::assertGreaterThan(0, $session->allocatedBytes());
    }

    public function test_session_reset_is_optional_and_keeps_session_usable(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"+":[{"var":"x"},1]}');
        $session = $engine->openSession();
        self::assertSame('2', $session->evaluate($rule, '{"x":1}'));
        $session->reset();
        self::assertSame('3', $session->evaluate($rule, '{"x":2}'));
    }

    public function test_parse_error_throws_ParseException(): void
    {
        $engine = new Engine();
        $this->expectException(ParseException::class);
        $engine->compile('not-json{{');
    }

    public function test_evaluate_error_throws_with_operator_and_path(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"throw":"boom"}');
        try {
            $rule->evaluate('{}');
            self::fail('expected EvaluateException');
        } catch (EvaluateException $ex) {
            self::assertSame('Thrown', $ex->errorType);
            self::assertNotNull($ex->pathJson);
            self::assertStringStartsWith('[', $ex->pathJson);
        }
    }

    public function test_templating_engine_constructs(): void
    {
        $engine = new Engine(templating: true);
        self::assertNotSame('', $engine->apply('{"+":[1,1]}', '{}'));
    }

    public function test_flagd_sem_ver_operator_is_available(): void
    {
        $engine = new Engine();
        self::assertSame('true', $engine->apply('{"sem_ver":["1.2.3","<","2.0.0"]}', '{}'));
    }

    public function test_traced_session_returns_result_and_steps(): void
    {
        $engine = new Engine();
        $session = $engine->openTracedSession();
        $run = $session->evaluate('{"+":[{"var":"x"},1]}', '{"x":41}');

        self::assertTrue($run->isSuccess());
        self::assertSame(42, $run->result);
        self::assertNotEmpty($run->steps);
        self::assertIsArray($run->expressionTree);
        self::assertNull($run->error);
    }

    public function test_traced_session_surfaces_runtime_error_in_payload(): void
    {
        $engine = new Engine();
        $session = $engine->openTracedSession();
        $run = $session->evaluate('{"throw":"boom"}', '{}');

        self::assertFalse($run->isSuccess());
        self::assertNotNull($run->error);
        self::assertIsArray($run->structuredError);
    }

    public function test_builder_registers_custom_operator(): void
    {
        $engine = Engine::builder()
            ->addOperator('double', function (string $argsJson): string {
                $arr = json_decode($argsJson, associative: true);
                return (string) ((int) $arr[0] * 2);
            })
            ->build();

        self::assertSame('42', $engine->apply('{"double":[21]}', '{}'));
    }

    public function test_builder_custom_operator_error_propagates(): void
    {
        $engine = Engine::builder()
            ->addOperator('boom', function (): string {
                throw new \RuntimeException('custom-failure');
            })
            ->build();

        try {
            $engine->apply('{"boom":[]}', '{}');
            self::fail('expected EvaluateException');
        } catch (EvaluateException $ex) {
            self::assertStringContainsString('custom-failure', $ex->getMessage());
        }
    }

    public function test_builder_set_config_json_strict_preset_takes_effect(): void
    {
        // Default config: null coerces to 0 and the sum evaluates.
        $lenient = new Engine();
        self::assertSame('1', $lenient->apply('{"+":[null,1]}', '{}'));

        // Strict preset: the same rule rejects the non-numeric null.
        $strict = Engine::builder()
            ->setConfigJson('{"preset":"strict"}')
            ->build();
        $this->expectException(EvaluateException::class);
        $strict->apply('{"+":[null,1]}', '{}');
    }

    public function test_builder_set_config_json_rejects_bad_input(): void
    {
        // Malformed JSON surfaces the parser's message.
        try {
            Engine::builder()->setConfigJson('not-json{{');
            self::fail('expected DatalogicException');
        } catch (DatalogicException $ex) {
            self::assertSame('ConfigurationError', $ex->errorType);
            self::assertNotSame('', $ex->getMessage());
        }

        // Unknown enum values fail loudly instead of being ignored.
        try {
            Engine::builder()->setConfigJson('{"preset":"bogus"}');
            self::fail('expected DatalogicException');
        } catch (DatalogicException $ex) {
            self::assertStringContainsString('bogus', $ex->getMessage());
        }
    }

    public function test_builder_set_config_json_chains_with_templating(): void
    {
        $engine = Engine::builder()
            ->withTemplating(true)
            ->setConfigJson('{"preset":"strict"}')
            ->build();
        self::assertSame('3', $engine->apply('{"+":[1,2]}', '{}'));

        $this->expectException(EvaluateException::class);
        $engine->apply('{"+":[null,1]}', '{}');
    }
}
